# First Contact Linear Regression

In this notebook, we aim to do linear regression! Our independent variables are throw-in length, height, angle, and location. Our dependent variable is the throw-in outcome (was the first contact sucessful). We conducted linear regression on each of the individual independent variables and also did a multiple linear regression with all the independent variables. 

In [5]:
# import libraries
from statsbombpy import sb
import numpy as np
import pandas as pd
import json
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn import linear_model
import statsmodels.api as sm

In [6]:
# obtaining the csv file with the event data from all matches
# creating a dataframe with this data
df_allData = pd.read_csv('/Users/rachelyang/Documents/GitHub/StatsBomb/full_data_sorted.csv')
df_allData

/var/folders/n4/wr2pjk454g5cxqdbd2qywndc0000gn/T/ipykernel_76237/3959350156.py:3: DtypeWarning: Columns (2,4,12,26,35,43,49,69,85,86,87,88,89,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111) have mixed types. Specify dtype option on import or set low_memory=False.
  df_allData = pd.read_csv('/Users/rachelyang/Documents/GitHub/StatsBomb/full_data_sorted.csv')


Unnamed: 0 ball_receipt_outcome ball_recovery_offensive  \
0                0                  NaN                     NaN   
1                1                  NaN                     NaN   
2                2                  NaN                     NaN   
3                3                  NaN                     NaN   
4                6                  NaN                     NaN   
...            ...                  ...                     ...   
192690        4814                  NaN                     NaN   
192691        4715                  NaN                     NaN   
192692        4815                  NaN                     NaN   
192693        4983                  NaN                     NaN   
192694        4984                  NaN                     NaN   

       ball_recovery_recovery_failure block_deflection carry_end_location  \
0                                 NaN              NaN                NaN   
1                                 NaN              NaN                NaN   
2                                 NaN              NaN                NaN   
3                                 NaN              NaN                NaN   
4                                 NaN              NaN                NaN   
...                               ...              ...                ...   
192690                            NaN              NaN                NaN   
192691                            NaN              NaN                NaN   
192692                            NaN              NaN                NaN   
192693                            NaN              NaN                NaN   
192694                            NaN              NaN                NaN   

       clearance_aerial_won clearance_body_part clearance_head  \
0                       NaN                 NaN            NaN   
1                       NaN                 NaN            NaN   
2                       NaN                 NaN            NaN   
3                       NaN                 NaN            NaN   
4                       NaN                 NaN            NaN   
...                     ...                 ...            ...   
192690                  NaN                 NaN            NaN   
192691                  NaN                 NaN            NaN   
192692                  NaN                 NaN            NaN   
192693                  NaN                 NaN            NaN   
192694                  NaN                 NaN            NaN   

       clearance_left_foot  ... shot_saved_to_post shot_redirect  \
0                      NaN  ...                NaN           NaN   
1                      NaN  ...                NaN           NaN   
2                      NaN  ...                NaN           NaN   
3                      NaN  ...                NaN           NaN   
4                      NaN  ...                NaN           NaN   
...                    ...  ...                ...           ...   
192690                 NaN  ...                NaN           NaN   
192691                 NaN  ...                NaN           NaN   
192692                 NaN  ...                NaN           NaN   
192693                 NaN  ...                NaN           NaN   
192694                 NaN  ...                NaN           NaN   

       pass_straight goalkeeper_shot_saved_off_target shot_saved_off_target  \
0                NaN                              NaN                   NaN   
1                NaN                              NaN                   NaN   
2                NaN                              NaN                   NaN   
3                NaN                              NaN                   NaN   
4                NaN                              NaN                   NaN   
...              ...                              ...                   ...   
192690           NaN                              NaN                   NaN   
192691           NaN                

### Cleaning and Filtering data

In [7]:
df_throwIn = df_allData.loc[df_allData['pass_type'] == 'Throw-in']
df_throwIn

Unnamed: 0 ball_receipt_outcome ball_recovery_offensive  \
12               8                  NaN                     NaN   
34              15                  NaN                     NaN   
38              16                  NaN                     NaN   
62              22                  NaN                     NaN   
71              25                  NaN                     NaN   
...            ...                  ...                     ...   
192468        1387                  NaN                     NaN   
192477        1390                  NaN                     NaN   
192501        1395                  NaN                     NaN   
192563        1415                  NaN                     NaN   
192585        1419                  NaN                     NaN   

       ball_recovery_recovery_failure block_deflection carry_end_location  \
12                                NaN              NaN                NaN   
34                                NaN              NaN                NaN   
38                                NaN              NaN                NaN   
62                                NaN              NaN                NaN   
71                                NaN              NaN                NaN   
...                               ...              ...                ...   
192468                            NaN              NaN                NaN   
192477                            NaN              NaN                NaN   
192501                            NaN              NaN                NaN   
192563                            NaN              NaN                NaN   
192585                            NaN              NaN                NaN   

       clearance_aerial_won clearance_body_part clearance_head  \
12                      NaN                 NaN            NaN   
34                      NaN                 NaN            NaN   
38                      NaN                 NaN            NaN   
62                      NaN                 NaN            NaN   
71                      NaN                 NaN            NaN   
...                     ...                 ...            ...   
192468                  NaN                 NaN            NaN   
192477                  NaN                 NaN            NaN   
192501                  NaN                 NaN            NaN   
192563                  NaN                 NaN            NaN   
192585                  NaN                 NaN            NaN   

       clearance_left_foot  ... shot_saved_to_post shot_redirect  \
12                     NaN  ...                NaN           NaN   
34                     NaN  ...                NaN           NaN   
38                     NaN  ...                NaN           NaN   
62                     NaN  ...                NaN           NaN   
71                     NaN  ...                NaN           NaN   
...                    ...  ...                ...           ...   
192468                 NaN  ...                NaN           NaN   
192477                 NaN  ...                NaN           NaN   
192501                 NaN  ...                NaN           NaN   
192563                 NaN  ...                NaN           NaN   
192585                 NaN  ...                NaN           NaN   

       pass_straight goalkeeper_shot_saved_off_target shot_saved_off_target  \
12               NaN                              NaN                   NaN   
34               NaN                              NaN                   NaN   
38               NaN                              NaN                   NaN   
62               NaN                              NaN                   NaN   
71               NaN                              NaN                   NaN   
...              ...                              ...                   ...   
192468           NaN                              NaN                   NaN   
192477           NaN                

In [8]:
# marking the complete first contact passes
df_throwIn.pass_outcome = df_throwIn.pass_outcome.fillna('Complete')
for row in df_throwIn.index:
    if (df_throwIn["pass_outcome"][row]== "Injury Clearance") or (df_throwIn["pass_outcome"][row]== "Unknown"):
        df_throwIn["pass_outcome"][row]="Incomplete"
df_throwIn["pass_outcome"]

/var/folders/n4/wr2pjk454g5cxqdbd2qywndc0000gn/T/ipykernel_76237/2721408823.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_throwIn.pass_outcome = df_throwIn.pass_outcome.fillna('Complete')
/var/folders/n4/wr2pjk454g5cxqdbd2qywndc0000gn/T/ipykernel_76237/2721408823.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_throwIn["pass_outcome"][row]="Incomplete"


12          Complete
34          Complete
38          Complete
62          Complete
71          Complete
             ...    
192468      Complete
192477    Incomplete
192501      Complete
192563      Complete
192585    Incomplete
Name: pass_outcome, Length: 1948, dtype: object

In [9]:
df_throwIn.drop(df_throwIn.columns.difference(['pass_outcome','location', "pass_height", "pass_angle", "pass_length"]), 1, inplace=True)
df_throwIn_sort = df_throwIn.reset_index()
df_throwIn_sort

/var/folders/n4/wr2pjk454g5cxqdbd2qywndc0000gn/T/ipykernel_76237/182920667.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  df_throwIn.drop(df_throwIn.columns.difference(['pass_outcome','location', "pass_height", "pass_angle", "pass_length"]), 1, inplace=True)
/var/folders/n4/wr2pjk454g5cxqdbd2qywndc0000gn/T/ipykernel_76237/182920667.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_throwIn.drop(df_throwIn.columns.difference(['pass_outcome','location', "pass_height", "pass_angle", "pass_length"]), 1, inplace=True)


index       location  pass_angle pass_height  pass_length pass_outcome
0         12    [43.1, 0.1]    2.876891    Low Pass    28.285154     Complete
1         34    [91.1, 0.1]    0.897825    Low Pass     8.823832     Complete
2         38    [83.0, 0.1]    1.229344    Low Pass    12.841339     Complete
3         62   [64.3, 80.0]   -2.664129    Low Pass    25.896140     Complete
4         71    [30.0, 0.1]    2.238189    Low Pass    27.627884     Complete
...      ...            ...         ...         ...          ...          ...
1943  192468  [106.8, 80.0]   -1.432344   High Pass    12.317873     Complete
1944  192477  [106.4, 80.0]   -1.482073   High Pass    32.728733   Incomplete
1945  192501    [36.2, 0.1]    2.527574    Low Pass    14.926487     Complete
1946  192563    [54.8, 0.1]    1.030918   High Pass    31.710093     Complete
1947  192585    [56.1, 0.1]    0.155665   High Pass    27.735357   Incomplete

[1948 rows x 6 columns]

In [10]:
for row in df_throwIn_sort.index:
    df_throwIn_sort["location"][row] = json.loads(df_throwIn_sort["location"][row])

/var/folders/n4/wr2pjk454g5cxqdbd2qywndc0000gn/T/ipykernel_76237/4050281909.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_throwIn_sort["location"][row] = json.loads(df_throwIn_sort["location"][row])


In [11]:
df_split_loc = pd.DataFrame(df_throwIn_sort['location'].tolist(), columns=['x_loc', 'y_loc'])
df_split_loc

x_loc  y_loc
0      43.1    0.1
1      91.1    0.1
2      83.0    0.1
3      64.3   80.0
4      30.0    0.1
...     ...    ...
1943  106.8   80.0
1944  106.4   80.0
1945   36.2    0.1
1946   54.8    0.1
1947   56.1    0.1

[1948 rows x 2 columns]

In [12]:
df_throwIn_sorted = df_throwIn_sort.drop(columns="location")
df_throwIn_final = df_throwIn_sorted.assign(x_loc = df_split_loc["x_loc"], y_loc = df_split_loc["y_loc"])
df_throwIn_final

index  pass_angle pass_height  pass_length pass_outcome  x_loc  y_loc
0         12    2.876891    Low Pass    28.285154     Complete   43.1    0.1
1         34    0.897825    Low Pass     8.823832     Complete   91.1    0.1
2         38    1.229344    Low Pass    12.841339     Complete   83.0    0.1
3         62   -2.664129    Low Pass    25.896140     Complete   64.3   80.0
4         71    2.238189    Low Pass    27.627884     Complete   30.0    0.1
...      ...         ...         ...          ...          ...    ...    ...
1943  192468   -1.432344   High Pass    12.317873     Complete  106.8   80.0
1944  192477   -1.482073   High Pass    32.728733   Incomplete  106.4   80.0
1945  192501    2.527574    Low Pass    14.926487     Complete   36.2    0.1
1946  192563    1.030918   High Pass    31.710093     Complete   54.8    0.1
1947  192585    0.155665   High Pass    27.735357   Incomplete   56.1    0.1

[1948 rows x 7 columns]

Since we are unable to use categorical data for our linear regression, we use dummy variables for both throw-in height and pass outcome. 

In [13]:
# convert "height" to dummy variable
df_throwIn_dummy = pd.get_dummies(df_throwIn_final, columns=['pass_height'], drop_first=True)
# high pass = 0, low pass = 1
df_throwIn_dummy.rename(columns={'pass_height_Low Pass':'pass_height'},inplace=True)
df_throwIn_dummy

index  pass_angle  pass_length pass_outcome  x_loc  y_loc  pass_height
0         12    2.876891    28.285154     Complete   43.1    0.1            1
1         34    0.897825     8.823832     Complete   91.1    0.1            1
2         38    1.229344    12.841339     Complete   83.0    0.1            1
3         62   -2.664129    25.896140     Complete   64.3   80.0            1
4         71    2.238189    27.627884     Complete   30.0    0.1            1
...      ...         ...          ...          ...    ...    ...          ...
1943  192468   -1.432344    12.317873     Complete  106.8   80.0            0
1944  192477   -1.482073    32.728733   Incomplete  106.4   80.0            0
1945  192501    2.527574    14.926487     Complete   36.2    0.1            1
1946  192563    1.030918    31.710093     Complete   54.8    0.1            0
1947  192585    0.155665    27.735357   Incomplete   56.1    0.1            0

[1948 rows x 7 columns]

In [14]:
#convert "outcome" to dummy variable
df_throwIn_dummy1 = pd.get_dummies(df_throwIn_dummy, columns=['pass_outcome'], drop_first=True)
# complete = 0, incomplete = 1
df_throwIn_dummy1.rename(columns={'pass_outcome_Incomplete':'pass_outcome'},inplace=True)
df_throwIn_dummy1.drop(columns="pass_outcome_Out",inplace=True)
df_throwIn_dummy1

index  pass_angle  pass_length  x_loc  y_loc  pass_height  pass_outcome
0         12    2.876891    28.285154   43.1    0.1            1             0
1         34    0.897825     8.823832   91.1    0.1            1             0
2         38    1.229344    12.841339   83.0    0.1            1             0
3         62   -2.664129    25.896140   64.3   80.0            1             0
4         71    2.238189    27.627884   30.0    0.1            1             0
...      ...         ...          ...    ...    ...          ...           ...
1943  192468   -1.432344    12.317873  106.8   80.0            0             0
1944  192477   -1.482073    32.728733  106.4   80.0            0             1
1945  192501    2.527574    14.926487   36.2    0.1            1             0
1946  192563    1.030918    31.710093   54.8    0.1            0             0
1947  192585    0.155665    27.735357   56.1    0.1            0             1

[1948 rows x 7 columns]

In [15]:
# let's see the data
df_throwIn_dummy1

index  pass_angle  pass_length  x_loc  y_loc  pass_height  pass_outcome
0         12    2.876891    28.285154   43.1    0.1            1             0
1         34    0.897825     8.823832   91.1    0.1            1             0
2         38    1.229344    12.841339   83.0    0.1            1             0
3         62   -2.664129    25.896140   64.3   80.0            1             0
4         71    2.238189    27.627884   30.0    0.1            1             0
...      ...         ...          ...    ...    ...          ...           ...
1943  192468   -1.432344    12.317873  106.8   80.0            0             0
1944  192477   -1.482073    32.728733  106.4   80.0            0             1
1945  192501    2.527574    14.926487   36.2    0.1            1             0
1946  192563    1.030918    31.710093   54.8    0.1            0             0
1947  192585    0.155665    27.735357   56.1    0.1            0             1

[1948 rows x 7 columns]

### Linear Regression with Throw-in Angle

In [16]:
# linear regression with angle
x_angle = df_throwIn_dummy1[['pass_angle']]
x_angle = sm.add_constant(x_angle)
np.asarray(x_angle)

array([[1.        , 2.876891  ],
       [1.        , 0.8978253 ],
       [1.        , 1.2293439 ],
       ...,
       [1.        , 2.5275736 ],
       [1.        , 1.0309176 ],
       [1.        , 0.15566467]])

In [17]:
y = df_throwIn_dummy1['pass_outcome']
y
np.asarray(y)

array([0, 0, 0, ..., 0, 0, 1], dtype=uint8)

In [18]:
model_angle = sm.OLS(y, x_angle)

In [19]:
results_angle = model_angle.fit()

In [20]:
angle_result = results_angle.summary()
print(angle_result)

                            OLS Regression Results                            
Dep. Variable:           pass_outcome   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.000
Method:                 Least Squares   F-statistic:                   0.09391
Date:                Mon, 01 Aug 2022   Prob (F-statistic):              0.759
Time:                        01:09:32   Log-Likelihood:                -746.32
No. Observations:                1948   AIC:                             1497.
Df Residuals:                    1946   BIC:                             1508.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.1478      0.008     18.364      0.0

In [21]:
print(f"coefficient of determination: {results_angle.rsquared}")

print(f"adjusted coefficient of determination: {results_angle.rsquared_adj}")

print(f"regression coefficients: {results_angle.params}")

coefficient of determination: 4.8253772105821824e-05
adjusted coefficient of determination: -0.0004655960460997477
regression coefficients: const         0.147791
pass_angle   -0.001390
dtype: float64


In [22]:
print(f"predicted response:\n{results_angle.fittedvalues}")

predicted response:
0       0.143792
1       0.146543
2       0.146082
3       0.151494
4       0.144680
          ...   
1943    0.149782
1944    0.149851
1945    0.144278
1946    0.146358
1947    0.147574
Length: 1948, dtype: float64


We expect that for every one radian increase in angle, there is a 0.0014 decrease in the probability of throw-in success. Based off of our r-squared value of 4.8253772105821824e-05, throw-in angle explains an extremely small proportion of the variation in throw-in success. Becuase the p-value is greater than 0.05, our data is not statistically significant. 
This is not the best result! Let's try throw-in length! (hopefully this one is better) 

### Linear Regression with Throw-in Length

In [23]:
# linear regression with length
x_length = df_throwIn_dummy1[['pass_length']]
x_length = sm.add_constant(x_length)
np.asarray(x_length)

array([[ 1.      , 28.285154],
       [ 1.      ,  8.823832],
       [ 1.      , 12.841339],
       ...,
       [ 1.      , 14.926487],
       [ 1.      , 31.710093],
       [ 1.      , 27.735357]])

In [24]:
# y stays the same
y = df_throwIn_dummy1['pass_outcome']
y
np.asarray(y)

array([0, 0, 0, ..., 0, 0, 1], dtype=uint8)

In [25]:
model_length = sm.OLS(y, x_length)
results_length = model_length.fit()
length_result = results_length.summary()
print(length_result)

                            OLS Regression Results                            
Dep. Variable:           pass_outcome   R-squared:                       0.087
Model:                            OLS   Adj. R-squared:                  0.086
Method:                 Least Squares   F-statistic:                     184.8
Date:                Mon, 01 Aug 2022   Prob (F-statistic):           2.83e-40
Time:                        01:09:33   Log-Likelihood:                -658.01
No. Observations:                1948   AIC:                             1320.
Df Residuals:                    1946   BIC:                             1331.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const          -0.0662      0.018     -3.776      

In [26]:
print(f"coefficient of determination: {results_length.rsquared}")

print(f"adjusted coefficient of determination: {results_length.rsquared_adj}")

print(f"regression coefficients: {results_length.params}")

coefficient of determination: 0.08672368558330845
adjusted coefficient of determination: 0.08625437606921971
regression coefficients: const         -0.066161
pass_length    0.011349
dtype: float64


In [27]:
print(f"predicted response:\n{results_length.fittedvalues}")

predicted response:
0       0.254846
1       0.033980
2       0.079575
3       0.227733
4       0.247386
          ...   
1943    0.073634
1944    0.305276
1945    0.103239
1946    0.293715
1947    0.248606
Length: 1948, dtype: float64


We expect that for every one yard increase in throw-in length, there is a 0.0113 increase in the probability of throw-in success. Based off of our r-squared value of 0.0867, throw-in length explains a small proportion of the variation in throw-in success. However, it is greater than throw-in angle. The p-value is less than 0.05, so our data is statistically significant! 
This is better, so now let's try height! 

### Linear Regression with Throw-in Height

In [28]:
# linear regression with height
x_height = df_throwIn_dummy1[['pass_height']]
x_height = sm.add_constant(x_height)
np.asarray(x_height)

array([[1., 1.],
       [1., 1.],
       [1., 1.],
       ...,
       [1., 1.],
       [1., 0.],
       [1., 0.]])

In [29]:
# y stays the same
y = df_throwIn_dummy1['pass_outcome']
y
np.asarray(y)

array([0, 0, 0, ..., 0, 0, 1], dtype=uint8)

In [30]:
model_height = sm.OLS(y, x_height)
results_height= model_height.fit()
height_result = results_height.summary()
print(height_result)

                            OLS Regression Results                            
Dep. Variable:           pass_outcome   R-squared:                       0.185
Model:                            OLS   Adj. R-squared:                  0.185
Method:                 Least Squares   F-statistic:                     443.1
Date:                Mon, 01 Aug 2022   Prob (F-statistic):           8.56e-89
Time:                        01:09:34   Log-Likelihood:                -546.55
No. Observations:                1948   AIC:                             1097.
Df Residuals:                    1946   BIC:                             1108.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const           0.4035      0.014     28.515      

In [31]:
print(f"coefficient of determination: {results_height.rsquared}")

print(f"adjusted coefficient of determination: {results_height.rsquared_adj}")

print(f"regression coefficients: {results_height.params}")

coefficient of determination: 0.18547473959656935
adjusted coefficient of determination: 0.18505617574230238
regression coefficients: const          0.403509
pass_height   -0.347063
dtype: float64


In [32]:
print(f"predicted response:\n{results_height.fittedvalues}")

predicted response:
0       0.056446
1       0.056446
2       0.056446
3       0.056446
4       0.056446
          ...   
1943    0.403509
1944    0.403509
1945    0.056446
1946    0.403509
1947    0.403509
Length: 1948, dtype: float64


We expect that for every one yard increase in height, there is a 0.3471 decrease in the probability of throw-in success. Based off of our r-squared value of 0.185, throw-in height explains a modest proportion of the variation in throw-in success, greater than both throw-in angle and throw-in length. The p-value is less than 0.05, so our data is statistically significant! 
Now, we move on to throw-in location! 

### Linear Regression with Throw-in Location

In [33]:
# linear regression with location
x_location = df_throwIn_dummy1[['x_loc']]
x_location = sm.add_constant(x_location)
np.asarray(x_location)

array([[ 1. , 43.1],
       [ 1. , 91.1],
       [ 1. , 83. ],
       ...,
       [ 1. , 36.2],
       [ 1. , 54.8],
       [ 1. , 56.1]])

In [34]:
# y stays the same
y = df_throwIn_dummy1['pass_outcome']
y
np.asarray(y)

array([0, 0, 0, ..., 0, 0, 1], dtype=uint8)

In [35]:
model_location = sm.OLS(y, x_location)
results_location = model_location.fit()
location_result = results_location.summary()
print(location_result)

                            OLS Regression Results                            
Dep. Variable:           pass_outcome   R-squared:                       0.011
Model:                            OLS   Adj. R-squared:                  0.010
Method:                 Least Squares   F-statistic:                     21.64
Date:                Mon, 01 Aug 2022   Prob (F-statistic):           3.51e-06
Time:                        01:09:35   Log-Likelihood:                -735.60
No. Observations:                1948   AIC:                             1475.
Df Residuals:                    1946   BIC:                             1486.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2320      0.020     11.726      0.0

In [36]:
print(f"coefficient of determination: {results_location.rsquared}")

print(f"adjusted coefficient of determination: {results_location.rsquared_adj}")

print(f"regression coefficients: {results_location.params}")

coefficient of determination: 0.010998041439191142
adjusted coefficient of determination: 0.010489818438902954
regression coefficients: const    0.23203
x_loc   -0.00128
dtype: float64


In [37]:
print(f"predicted response:\n{results_location.fittedvalues}")

predicted response:
0       0.176878
1       0.115455
2       0.125820
3       0.149749
4       0.193641
          ...   
1943    0.095365
1944    0.095876
1945    0.185707
1946    0.161906
1947    0.160242
Length: 1948, dtype: float64


We expect that for every one yard along the side of the field, there is a 0.0013 decrease in the probability of throw-in success. Based off of our r-squared value of 0.011, throw-in location explains a rather small proportion of the variation in throw-in success. It is still greater than throw-in angle. 
Since our individual linear regressions don't seem to be the best, let's try multiple linear regression!! 

### Multiple Linear Regression

After realizing that linear regression on independent variables by themselves was not going to work... we decided to give multiple linear regression a try! Hopefully, this works better!!!  

In [38]:
X = df_throwIn_dummy1[['pass_height', 'pass_angle', 'pass_length', 'x_loc']]
Y = df_throwIn_dummy1['pass_outcome']

In [39]:
from sklearn import linear_model
regr = linear_model.LinearRegression()
regr.fit(X, y)

LinearRegression()

In [40]:
print(regr.coef_)

[-0.29711107  0.00271456  0.00537342 -0.00030429]


In [43]:
print('Intercept: \n', regr.intercept_)
print('Coefficients: \n', regr.coef_)

Intercept: 
 0.2855085222526844
Coefficients: 
 [-0.29711107  0.00271456  0.00537342 -0.00030429]


In [42]:
# with statsmodels
x = sm.add_constant(X) # adding a constant
 
model = sm.OLS(Y, x).fit()
predictions = model.predict(x) 
 
print_model = model.summary()
print(print_model)

                            OLS Regression Results                            
Dep. Variable:           pass_outcome   R-squared:                       0.202
Model:                            OLS   Adj. R-squared:                  0.201
Method:                 Least Squares   F-statistic:                     123.2
Date:                Mon, 01 Aug 2022   Prob (F-statistic):           7.94e-94
Time:                        01:13:16   Log-Likelihood:                -526.15
No. Observations:                1948   AIC:                             1062.
Df Residuals:                    1943   BIC:                             1090.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const           0.2855      0.030      9.573      

A multiple regression will use all of our variables to predict throw-in outcome. Upon conducting the regression, we find that pass height and pass angle are two variables whose relationship with throw-in success is statistically significant (p-value ≤ 0.05). We expect that for each 1 yard increase in pass height, there is a 0.2971 decrease in the probability of throw-in success. And, for each 1 yard increase in pass length, we predict a 0.0054 increase in the probability of throw-in success. Pass angle and pass location were not statistically significant variables in predicting throw-in success. Overall, our model’s r-squared is 0.202, which we believe to be quite high for real-world data.

We realize that linear regressions may not be the best model for our data. We are now searching for new models that may be a better fit. We also want to research more variables that may influence the success rate of a throw-in. The data shows that the coefficient constant is 0.202 which indicates that there are other variables that have potential to influence the success of a throw-in! 